content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,243 @@
+  else
+    "#{title} - #{last_name}, #{middle_name}, #{first_name} (#{publish_year})"
+  end
+end
+```
+#### After refactoring
+```rb
+def summary
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+```
+Obviously, the code quality has a huge improvement after refactoring.
+
+## Loose coupling
+Before we finish this refactoring and send the PR to someone, let's check the code again, does it clean enough?
+
+The code seems short and simple, but this method still depends on the author class. In other words, the `summary` method needs to know the author object has a `full_name` method, we could say the book class still couple with author class. In this case, maybe it's hard to notice the cost of coupling, let's imagine a few years later, the `summary` method might become more complicated:
+
+#### Few years later
+```rb
+def summary
+  # 10 lines of code you don't want to read.
+  middle_result = "#{some_author_title_helper(author)} - #{author.full_name}"
+  # another 10 lines of code
+  return final_result
+end
+```
+The disadvantage of coupling is that when one class needs to change, you need to change others as well. For example, one day we need to rename the full name method or add more parameters to it, we need to modify the book class at the same time, which we don't like.
+
+The bad news is, in the real world it's almost impossible to remove all the coupling between each object, but we can at least reduce the impact by segregating and highlighting the dependence. 
+
+```rb
+def summary
+  "#{title} - #{author_name} (#{publish_year})"
+end
+
+def author_name
+  author.full_name
+end
+```
+After this refactoring, the `summary` method 100% calling from the book object itself. The book class still has the coupling with the author class, but it's obvious and easy to maintain. If we only consider the example case, you might wonder is that a little bit over design? Well, I would say that a good pattern is never too early. Now we can push the PR confidently and enjoy the holiday.
+
+## Conclusion
+Good code should be small and simple. Of course, the parameter is not the only place we can place our effort in to reducing its size. During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe this line of thinking would help you perform well in development speed and product stability in the long term.
+
+I hope this small example can convince you this refactoring skill can benefit your coding life. If not, please wait for my next blog and give me another chance. See you next time.
+
+[1]: https://en.wikipedia.org/wiki/Duck_typing
+[2]: https://en.wikipedia.org/wiki/Law_of_Demeter
+[3]: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle
```

Comment:
I don't have time to review all the grammar but it needs more proofreading.

The content itself is good and generally important. Some general comments:

1. We should probably mention this is a particular case of a more general code-smell called [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession). Primitive Obsession was coined by Martin Fowler, as far as I can tell.
2. We should also mention the [Whole Value idiom](http://fit.c2.com/wiki.cgi?WholeValue) promoted by Ward Cunningham even earlier. This provides the philosophical reasoning behind "why we merge first name and last name" - because these things represents a different value.
3. The `Author` example is not great IMHO. The actual value we really have in this argument is a person's _name_, not an author. A proper `Author` class would have things like birthday, country, biography and profile image. I think a better example would be for a `PersonName` which can be reused in parameters to functions, the Author class and even the User class. Another example could be the canonical [Whole Value](http://fit.c2.com/wiki.cgi?WholeValue): A pair of currency type and value (`15.33` and `:USD`).